hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerProtocolServerSideTranslatorPB.java

```diff
@@ -135,7 +141,9 @@ public OMResponse submitRequest(RpcController controller,
       OMRequest request) throws ServiceException {
     OMRequest validatedRequest;
     try {
-      validatedRequest = requestValidations.validateRequest(request);
+      OMRequest requestWithBucketId = associateBucketIdWithRequest(request);
```

Comment:
We had considered adding the bucket ID in the pre-process validators, but the issue is that we don't have validators for all kinds of requests. We'll end up validating bucket IDs only for requests that have upgrade validations in place.

But I am with you on the idea of reducing the switch-case-like code `associateBucketIdWithRequest`. We're also debating on a way to make sure that when new request classes are added that we raise some test errors if the person forgets to add the related logic to validate the bucket ID.

Reply:
> but the issue is that we don't have validators for all kinds of requests.

Ah ok this jogged my memory. This type of race actually causes two different issues:
1. An old client could mistakenly reach an FSO bucket because the validator read stale bucket layout info when admitting the request.
2. The type Ratis request (FSO or OBS specific) could mismatch the actual type of the bucket when the request is applied.

For issue 1, the requests will have request validators blocking client requests. In these validators we can add the bucket ID to the request.
For issue 2, the requests must go through the `BucketLayoutAwareOMKeyRequestFactory`. In this class's `createRequest` method, we could add the bucket ID to the request before we call `getBucketLayout`.

Does this seem like a workable solution?